### PR TITLE
librados: new style (sharded) object listing (#9963)

### PR DESCRIPTION
**Pull request by jcsp**
This is a rebase of what was wip-9963-sage

### CODE REVIEW

File: src/include/rados/librados.h [hunk 1]
```diff
@@ -931,6 +950,41 @@ CEPH_RADOS_API int rados_nobjects_list_next(rados_list_ctx_t ctx,
...
+    const size_t result_size,
+    rados_enumerate_item *results);
+
+CEPH_RADOS_API void rados_enumerate_objects_split(rados_ioctx_t io,
+    const rados_enumerate_cursor start,
+    const rados_enumerate_cursor finish,
+    const size_t n,
+    const size_t m,
+    rados_enumerate_cursor *split_start,
+    rados_enumerate_cursor *split_finish);
```
Comment (liewegas) [inline src/include/rados/librados.h]: can this get a comment documetning it?  i would expect a single output argument (the cursor at the approx middle/split point)

File: src/include/rados/librados.h [hunk 2]
```diff
@@ -190,6 +190,25 @@ typedef void *rados_ioctx_t;
...
 
 /**
+ * @typedef rados_enumerate_cursor
+ *
+ * The cursor used with rados_enumerate_objects
+ * and accompanying methods.
+ */
+typedef void * rados_enumerate_cursor;
+
+typedef struct rados_enumerate_item {
```
Comment (liewegas) [inline src/include/rados/librados.h]: rados_object_item

File: src/include/rados/librados.h [hunk 3]
```diff
@@ -190,6 +190,25 @@ typedef void *rados_ioctx_t;
 typedef void *rados_list_ctx_t;
 
 /**
+ * @typedef rados_enumerate_cursor
+ *
+ * The cursor used with rados_enumerate_objects
+ * and accompanying methods.
+ */
+typedef void * rados_enumerate_cursor;
```
Comment (liewegas) [inline src/include/rados/librados.h]: rados_object_list_cursor

File: src/librados/librados.cc [hunk 4]
```diff
@@ -3536,6 +3537,113 @@ extern "C" int rados_exec(rados_ioctx_t io, const char *o, const char *cls, cons
   return ret;
 }
 
+extern "C" rados_enumerate_cursor rados_enumerate_objects_begin(rados_ioctx_t io)
```
Comment (liewegas) [inline src/librados/librados.cc]: rados_object_list_begin

File: src/include/rados/librados.h (hunk 1 above)
Comment (jcsp) [inline src/include/rados/librados.h]: It took me a minute to remember so docstring definitely needed...

The dual output arguments get you a range from somewhere within the input start/finish, basically (chunk n of m).  Maybe "slice" is a better name than "split" here

File: src/librados/librados.cc [hunk 5]
```diff
@@ -3536,6 +3537,113 @@ extern "C" int rados_exec(rados_ioctx_t io, const char *o, const char *cls, cons
...
+{
+  librados::IoCtxImpl *ctx = (librados::IoCtxImpl *)io;
+
+  hobject_t *result = new hobject_t(ctx->objecter->enumerate_objects_begin());
+  return (rados_enumerate_cursor)result;
+}
+
+extern "C" rados_enumerate_cursor rados_enumerate_objects_end(rados_ioctx_t io)
```
Comment (liewegas) [inline src/librados/librados.cc]: rados_object_{begin,end}

**Comment by liewegas** [comment]
for the C++ API, object_list_{begin,end} is awkward but consistent with the C API and with other list methods...

File: src/include/rados/librados.h (hunk 1 above)
Comment (liewegas) [inline src/include/rados/librados.h]: Oh, right.  Yep, that makes sense!

Comment (liewegas) [inline src/include/rados/librados.h]: Slice sounds good to me too

**Comment by jcsp** [comment]
Updated with C++ bits.

**Comment by jcsp** [comment]
The reason I didn't do an STL-style iterator thing is that it's a pretty weird API to hide network IO like that, and exposing the "give me the next N + return an error code" interface adds very little burden to the caller.

**Comment by liewegas** [comment]
Agreed.. I found the STL-style iterator annoying anyway.

File: src/test/librados/list.cc [hunk 6]
```diff
@@ -665,5 +668,219 @@ TEST_F(LibRadosListECPP, ListObjectsStartPP) {
...
+  }
+
+  std::set<std::string> saw_obj;
+  ObjectCursor c = ioctx.object_list_begin();
+  ObjectCursor end = ioctx.object_list_end();
+  while(!ioctx.object_list_is_end(c))
+  {
+    std::vector<ObjectItem> result;
+    int r = ioctx.object_list(c, end, 12, &result, &c);
+    ASSERT_GE(r, 0);
```
Comment (liewegas) [inline src/test/librados/list.cc]: could ASSERT_EQ(r, result.size()) here too

File: src/include/rados/librados.h [hunk 7]
```diff
@@ -190,6 +190,25 @@ typedef void *rados_ioctx_t;
...
 
 /**
+ * @typedef rados_object_list_cursor
+ *
+ * The cursor used with rados_enumerate_objects
+ * and accompanying methods.
+ */
+typedef void * rados_object_list_cursor;
+
+typedef struct rados_object_item {
```
Comment (liewegas) [inline src/include/rados/librados.h]: On second thought this should probably be rados_object_list_item, yeah?

**Comment by jcsp** [comment]
Updated

**Comment by jcsp** [comment]
loic-bot's failure is weird, because my test branch (including this) is building on the gitbuilders, and when I try an autotools build locally it works as well.

It's failing like this:
<pre>
  CXXLD    unittest_erasure_code_plugin_shec
  CXXLD    unittest_erasure_code_example
  CXXLD    unittest_librados
  CXXLD    unittest_librados_config
  CXXLD    unittest_journal
  CXXLD    unittest_rbd_replay
  CXXLD    unittest_encoding
  CXXLD    unittest_base64
  CXXLD    unittest_run_cmd
  CXXLD    unittest_simple_spin
  CXXLD    unittest_libcephfs_config
  CXXLD    unittest_mon_moncap
  CXXLD    unittest_mon_pgmap
  CXXLD    unittest_ecbackend
  CXXLD    unittest_osdscrub
  CXXLD    unittest_pglog
  CXXLD    unittest_hitset
  CXXLD    unittest_osd_osdcap
  CXXLD    unittest_chain_xattr
  CXXLD    unittest_lfnindex
  CXXLD    unittest_mds_authcap
  CXXLD    unittest_addrs
  CXXLD    unittest_blkdev
  CXXLD    unittest_bloom_filter
  CXXLD    unittest_histogram
  CXXLD    unittest_prioritized_queue
  CXXLD    unittest_str_map
./.libs/libradostest.a(libradostest_la-test.o): In function `wait_for_healthy(void**)':
/home/jenkins/workspace/ceph/LABELS/ceph-centos-7-jenkins/src/test/librados/test.cc:67: undefined reference to `rados_buffer_free'
collect2: error: ld returned 1 exit status
  CXXLD    unittest_sharedptr_registry
make[4]: *** [unittest_journal] Error 1
make[4]: *** Waiting for unfinished jobs....
make[4]: Leaving directory `/home/jenkins/workspace/ceph/LABELS/ceph-centos-7-jenkins/src'
make[3]: *** [check-am] Error 2
make[3]: Leaving directory `/home/jenkins/workspace/ceph/LABELS/ceph-centos-7-jenkins/src'
make[2]: *** [check-recursive] Error 1
make[2]: Leaving directory `/home/jenkins/workspace/ceph/LABELS/ceph-centos-7-jenkins/src'
make[1]: *** [check] Error 2
make[1]: Leaving directory `/home/jenkins/workspace/ceph/LABELS/ceph-centos-7-jenkins/src'
make: *** [check-recursive] Error 1
</pre>

It also makes no sense that rados_buffer_free would be missing from anything that was linking with libradostest, because that lib was already using rados_mon_command et al.

**Comment by jcsp** [comment]
It's unittest_journal, which doesn't correctly link with librados

**Comment by dachary** [comment]
@jcsp it must be something specific to CentOS 7 (since it fails consistently) and it should be easy to reproduce on a virgin CentOS 7.

File: src/test/Makefile-client.am [hunk 8]
```diff
@@ -29,7 +29,7 @@ libradostest_la_SOURCES = \
 	test/librados/TestCase.cc
 noinst_LTLIBRARIES += libradostest.la
 libradostest_la_CXXFLAGS = $(UNITTEST_CXXFLAGS)
-RADOS_TEST_LDADD = libradostest.la
+RADOS_TEST_LDADD = libradostest.la $(LIBCOMMON) $(CRYPTO_LIBS)
```
Comment (dachary) [inline src/test/Makefile-client.am]: I bet the link failure is a tasty autotools side effect of this addition :-(

**Comment by jcsp** [comment]
Updated with (hopefully) fixed autotools build of unit tests

**Comment by jcsp** [comment]
See a run with no failures attributable to this PR here: http://pulpito.ceph.com/jspray-2015-11-30_02:06:10-fs-wip-testing-jcsp---basic-multi/

This is good to go from my POV, I'll hook DataScan into it in a separate PR.

**Comment by jcsp** [comment]
Oops, this lost the update to unit test linking when I was adding sage's commits.  repushed.

**Comment by liewegas** [comment]
2015-12-03T20:09:44.741 DEBUG:tasks.ceph.mon.d:waiting for process to exit
2015-12-03T20:09:49.011 INFO:tasks.workunit.client.0.plana17.stdout:test/librados/list.cc:731: Failure
2015-12-03T20:09:49.011 INFO:tasks.workunit.client.0.plana17.stdout:Value of: err_str.empty()
2015-12-03T20:09:49.012 INFO:tasks.workunit.client.0.plana17.stdout:  Actual: false
2015-12-03T20:09:49.012 INFO:tasks.workunit.client.0.plana17.stdout:Expected: true
2015-12-03T20:09:49.018 INFO:tasks.workunit.client.0.plana17.stderr:Segmentation fault (core dumped)

from /a/sage-2015-12-03_13:46:35-rados-wip-sage-testing2---basic-multi/1167298

**Comment by jcsp** [comment]
Huh -- I didn't realise we ran unit tests on a thrashing cluster.  In this instance it took about 120 seconds for the test's PGs to get created, which exceeded the 30s timeout in wait_for_healthy.  Guess I'll crank up that timeout.

**Comment by jcsp** [comment]
Pushed update for timeouts.

**Comment by jcsp** [comment]
ping @liewegas

**Comment by liewegas** [comment]
@jcsp yeah, just waiting for the lab to come back alive so i can run it through again.

**Pull request merged**